### PR TITLE
Feature/qos service api

### PR DESCRIPTION
**Pull request by gsheasby**
**Goals (and why)**: Work in progress - add a Quality of Service service, and call it when Cassandra.Client makes requests to Cassandra. Note that this is very basic so far; I'm throwing up the PR to solicit feedback on design decisions so far and future direction.

We would like to keep the QoS service implementation internal to OSS AtlasDB, and introduce a magic layer internally as a thin wrapper, similar to our approach for TimeLock. This will make our testing easier and more robust.

**Implementation Description (bullets)**:
- [x] Create basic QoS service
- [x] Create basic QoS client that calls that service
- [x] Wrap `Cassandra.Client` with `CassandraClient`, and call `qosClient.checkLimit` when performing some operations
- [x] Service hands out MAX_VALUE "credits"; client uses up 1 "credit" per operation

Next:
- [x] `CassandraClient` should also call `checkLimit` for CQL query execution
- [x] Client periodically calls `service.getLimit`; this refreshes the "credits"
- [x] Centralise service (should be one service per stack, and one client per product node)
- [x] Client reports usage metrics to server
- [x] Server can configure different limits per client

Future (but still before merge):
- [x] Different amounts of credits are used up depending on the complexity of the request
- [x] Implement soft limiting
- [x] Server updates limits based on Cassandra load
- [x] Figure out exception handling

**Concerns (what feedback would you like?)**: Design decisions; are the next/future steps reasonable?

**Where should we start reviewing?**: CassandraClient

**Priority (whenever / two weeks / yesterday)**: **do not merge** - but will be iterating quickly on this over the next week.

---
This change is [Reviewable](https://reviewable.io/reviews/palantir/atlasdb/2629)

### CODE REVIEW

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClient.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+    }
+
+    @Override
+    public List<KeySlice> get_range_slices(ColumnParent column_parent, SlicePredicate predicate, KeyRange range,
+            ConsistencyLevel consistency_level)
+            throws InvalidRequestException, UnavailableException, TimedOutException, TException {
+        return checkLimitAndCall(() -> super.get_range_slices(column_parent, predicate, range, consistency_level));
+    }
+
+    private <T> T checkLimitAndCall(Callable<T> callable) {
```
Comment (nziebart) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClient.java]: what's the purpose of accepting a callable here? why not just `checkLimit()` before each call?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClient.java [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+        try {
+            qosClient.checkLimit();
+            return callable.call();
+        } catch (Exception ex) {
+            throw Throwables.throwUncheckedException(ex);
```
Comment (nziebart) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClient.java]: we probably dont want to wrap these exceptions

Comment (hsaraogi) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClient.java]: Fixed.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClient.java (hunk 1 above)
Comment (hsaraogi) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/CassandraClient.java]: Got rid of this method.

File: atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLockTestTools.java [hunk 3]
```diff
@@ -51,20 +50,21 @@ public CqlResult setLocksTableValue(long lockId, int heartbeatCount) throws TExc
...
+            return runCqlQuery(truncateQuery, client, ConsistencyLevel.ALL);
         });
     }
 
     public CqlResult readLocksTable() throws TException {
         return clientPool.run(client -> {
             String lockRowName = getHexEncodedBytes(CassandraConstants.GLOBAL_DDL_LOCK_ROW_NAME);
             String lockColName = getHexEncodedBytes(CassandraConstants.GLOBAL_DDL_LOCK_COLUMN_NAME);
-            String selectCql = String.format(
+            CqlQuery selectCql = new CqlQuery(
```
Comment (sandorw) [inline atlasdb-cassandra-integration-tests/src/test/java/com/palantir/atlasdb/keyvalue/cassandra/SchemaMutationLockTestTools.java]: Is this related to the QoS service change?

File: qos-service-api/src/main/java/com/palantir/atlasdb/qos/QosService.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+
+import com.palantir.logsafe.Safe;
+
+@Path("/qos")
+public interface QosService {
+    @Path("{client: [a-zA-Z0-9_-]+}/get-limit")
+    @POST
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    long getLimit(@Safe @PathParam("client") String client);
```
Comment (sandorw) [inline qos-service-api/src/main/java/com/palantir/atlasdb/qos/QosService.java]: Our meters are tracking #s of requests and bytes read/written. Which limit is this meant to represent - one of them or some combined limit across all of them?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/qos/ThriftObjectSizeUtils.java [hunk 5]
```diff
@@ -0,0 +1,264 @@
...
+                        + getCollectionSize(rowResult.getValue(),
+                        ThriftObjectSizeUtils::getColumnOrSuperColumnSize));
+    }
+
+    public static long getApproximateSizeOfKeySlices(List<KeySlice> slices) {
+        return getCollectionSize(slices, ThriftObjectSizeUtils::getKeySliceSize);
+    }
+
+    public static long getCasByteCount(List<Column> updates) {
+        // TODO(nziebart): CAS actually writes more bytes than this, because the associated Paxos negotations must
```
Comment (hsaraogi) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/qos/ThriftObjectSizeUtils.java]: @nziebart Given that CAS is not going through the rate limiter, we might be able to skip this TODO?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/qos/ThriftQueryWeighers.java [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+        return zeroEstimate
+                ? readWeigherWithZeroEstimate(ThriftObjectSizeUtils::getColumnOrSuperColumnSize, ignored -> 1, 1)
+                : readWeigher(ThriftObjectSizeUtils::getColumnOrSuperColumnSize, ignored -> 1, 1);
+    }
+
+    static final QosClient.QueryWeigher<CqlResult> EXECUTE_CQL3_QUERY =
+            // TODO(nziebart): we need to inspect the schema to see how many rows there are - a CQL row is NOT a
+            // partition. rows here will depend on the type of query executed in CqlExecutor: either (column, ts) pairs,
+            // or (key, column, ts) triplets
+            // Currently, transaction or metadata table queries dont use the CQL executor,
```
Comment (hsaraogi) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/qos/ThriftQueryWeighers.java]: I added a zero estimation for other transaction or metadata queries. Since its not using the cql executor we can get away with this for now.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/qos/ThriftQueryWeighers.java [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+                : readWeigher(ThriftObjectSizeUtils::getApproximateSizeOfKeySlices, List::size, keyRange.count);
+    }
+
+    static QosClient.QueryWeigher<ColumnOrSuperColumn> get(boolean zeroEstimate) {
+        return zeroEstimate
+                ? readWeigherWithZeroEstimate(ThriftObjectSizeUtils::getColumnOrSuperColumnSize, ignored -> 1, 1)
+                : readWeigher(ThriftObjectSizeUtils::getColumnOrSuperColumnSize, ignored -> 1, 1);
+    }
+
+    static final QosClient.QueryWeigher<CqlResult> EXECUTE_CQL3_QUERY =
```
Comment (hsaraogi) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/qos/ThriftQueryWeighers.java]: `EXECUTE_CQL3_QUERY` is a misleading name for the weigher. Can we estimate the number of rows from the `CqlResult`, each CqlRow (partition) has a list of columns (each of which has a ts) - is `CqlResult.getRows().stream().mapToInt(row -> row.getColumns.size()).sum()` the estimate here?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/qos/ThriftQueryWeighers.java [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+                return ImmutableQueryWeight.builder()
+                        .from(estimate())
+                        .timeTakenNanos(timeTakenNanos)
+                        .build();
+            }
+        };
+    }
+
+    // TODO(nziebart): we really shouldn't be needing to catch exceptions here
+    private static long safeGetNumBytesOrDefault(Supplier<Long> numBytes) {
```
Comment (hsaraogi) [inline atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/qos/ThriftQueryWeighers.java]: Most kinds of thrift objects with different fields filled in or empty have been tested, but I wouldnt stay the tests are comprehensive and cover all kinds of objects. Do we want to add more tests, is there some way to just generate a bunch of objects and ensure that the size calculating method doesn't throw?

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/qos/ThriftQueryWeighers.java (hunk 7 above)
Comment (nziebart) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/qos/ThriftQueryWeighers.java]: a `CqlRow` is not the same as a cassandra row. e.g., if i `select (key, column1, column2)` i will get 1 `CqlRow` per column, but there's a difference between having 1000 columns in a single row vs 1000 rows with a single column

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/qos/ThriftQueryWeighers.java (hunk 8 above)
Comment (nziebart) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/qos/ThriftQueryWeighers.java]: i would think that the cassandra ETE tests would catch these errors. but i'm ok with leaving this for now, and removing it fairly soon after we can verify that this error doesn't occur in the field.

File: atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/qos/ThriftObjectSizeUtils.java (hunk 5 above)
Comment (nziebart) [reply atlasdb-cassandra/src/main/java/com/palantir/atlasdb/keyvalue/cassandra/qos/ThriftObjectSizeUtils.java]: yeah, we can leave this alone for now. ideally we still record metrics for CAS, we just don't actually apply rate limiting to it. but we don't need to block on this
